File: imposc/lib/periodic.ex
import ImposcUtils
import SystemParameters

defmodule OneNParams do
  @moduledoc """
  Structure for parameters for analysing single-impact (1, n) periodic orbits as the obstacle offset sigma varies
  """
  
  @doc """
  `:omega`: the forcing frequency
  `:r`: the coefficient of restitution at an impact
  `:gamma`: the coefficient of the forcing term in the equation for the displacement between impacts
  `:gamma2`: square of the coefficient of the forcing term in the equation for the displacement between impacts
  `:r_minus`: (1 - r)/omega - a coefficient used in the analysis
  `:n`: the number of forcing cycles between each impact
  `:cs`: a coefficient used in the analysis
  `:sigma_s`: The positive intercept with the sigma-axis
  `:period`: The period between impacts
  """

  defstruct omega: 2, r: 0.8, gamma: 1/3.0, gamma2: 1/9.0, r_minus: 0.1, n: 1, cs: 0, sigma_s: 0, period: :math.pi

  defp derive_cs(cn, _sn, _r) when cn == 1 do
    0
  end

  defp derive_cs(cn, sn, r) do
   sn*(1+r)/(1-cn)
  end

  @doc """
  Derives a `:OneNParams` instance from a minimal parameter set

  `:omega`: the forcing frequency
  `:r`: the coefficient of restitution at an impact
  `:n`: the number of forcing cycles between impacts in a single-impact periodic orbit
  """

  @spec derive(float, float, integer) :: OneNParams
  def derive(omega, r, n) do
    period = n * ImposcUtils.forcing_period(omega)
    cn = :math.cos(period)
    sn = :math.sin(period)
    gamma = ImposcUtils.gamma(omega)

    result = %OneNParams{omega: omega, r: r, gamma2: gamma * gamma, r_minus: (1-r)/omega, cs: derive_cs(cn, sn, r)}
    result = %OneNParams{result | sigma_s: :math.sqrt(result.gamma2*(1 + :math.pow(result.cs/result.r_minus, 2)))}
    result = %OneNParams{result | period: period, gamma: gamma}
    result
  end

  @spec discriminant(number, OneNParams) :: number
  defp discriminant(sigma, %OneNParams{} = params) do
    # The discriminant of the quadratic equation for the impact velocity for a given value of the obstacle offset.
    4 * (params.gamma2*params.cs*params.cs - (sigma  * sigma - params.gamma2) * params.r_minus * params.r_minus)
  end

  @spec velocities_for_discr(number, number, OneNParams) :: [nil | float, ...]
  #  Solves the quadratic equation to return the velocities for candidate (1, n) orbits for a given obstacle offset and
  #  discriminant. Depending on the value of the discriminant, there will be either zero, one (in the case of a double
  #  root) or two such velocities. Unphysical or negative-velocity orbits are not filtered out.
  defp velocities_for_discr(_sigma, discriminant, %OneNParams{} = _params) when discriminant < 0 do
    # Complex roots
    [nil, nil]
  end

  defp velocities_for_discr(sigma, discriminant, %OneNParams{} = params) do
    # Real roots
    intercept = -2 * params.cs * sigma
    divisor = params.cs * params.cs + params.r_minus * params.r_minus

    vs = intercept/divisor

    d = :math.sqrt(discriminant) / divisor

    [vs + d, vs - d]
  end

  @spec phase_for_velocity( float | nil, float, OneNParams) :: float
  # Returns the phase corresponding to a solution of the quadratic equation for the velocity of a (1, n) orbit.
  defp phase_for_velocity(nil, _sigma, %OneNParams{} = _params) do
    # Return `nil` for a `nil` velocity
    nil
  end

  defp phase_for_velocity(velocity, sigma, %OneNParams{} = params) do
    arg = (sigma + params.cs * velocity / 2.0) / params.gamma

    angle = :math.acos(arg)

    # Check for phase correction. sin(angle) should have opposite sign to v/gamma
    s = :math.sin(angle)
    vg = velocity / params.gamma
    angle =
      cond do
        (vg > 0 and s > 0) or (vg < 0 and s < 0) -> 2 * :math.pi - angle
        true -> angle
      end

    ImposcUtils.phi(angle / params.omega, params.omega)
  end

  @doc """
  Returns an `:ImpactPoint` corresponding to a solution of the quadratic equation for the velocity of a (1, n)
  orbit.

  `:velocity`: impact velocity for a candidate (1, n) orbit
  `:sigma`: the obstacle offset
  `:params`: parameters held fixed as the offset varies for a specified (1, n) orbit
  """

  def point_for_velocity(nil, _sigma, %OneNParams{} = _params) do
    nil
  end

  def point_for_velocity(velocity, sigma, %OneNParams{} = params) do
    phase_for_velocity(velocity, sigma, params) |> (&%ImpactPoint{phi: &1, v: velocity, t: &1}).()
  end

  @doc """
  Returns impact velocities corresponding to physical (1, n) orbits for a given obstacle offset. There will be
  either zero, one (in the case of a double root) or two such velocities.

  `:sigma`: the obstacle offset
  `:params`: parameters held fixed as the offset varies for a specified (1, n) orbit
  """

  @spec velocities(number, OneNParams) :: {nil | float, nil | float}
  def velocities(sigma, %OneNParams{} = params) do
#    IO.puts("#{sigma}, #{params.sigma_s}")
    velocities_for_discr(sigma, discriminant(sigma, params), params) |> Enum.map(&nullify_unphysical(&1, sigma, params)) |> List.to_tuple
  end

  @doc """
  Returns points on the impact surface corresponding to physical (1, n) orbits for a given obstacle offset. There
  will be either zero, one (in the case of a double root) or two such points.

  `:sigma`: the obstacle offset
  `:params`: parameters held fixed as the offset varies for a specified (1, n) orbit
  """

  @spec orbits(number, OneNParams) :: [any]
  def orbits(sigma, %OneNParams{} = params) do
    velocities(sigma, params) |> Tuple.to_list |> Enum.map(& point_for_velocity(&1, sigma, params))
  end

  @doc """
  Filters out velocities for unphysical (1, n) orbits.

  `:velocity`: impact velocity for a candidate (1, n) orbit
  `:sigma`: the obstacle offset
  `:params`: parameters held fixed as the offset varies for a specified (1, n) orbit

  Returns `:velocity` if physical, `:nil` if unphysical.
  """

  @spec nullify_unphysical(any, any, OneNParams) :: any
  def nullify_unphysical(velocity, sigma, %OneNParams{} = params) do
    if is_physical?(velocity, sigma, params) do
      velocity
    else
      nil
    end
  end

  @doc """
  Checks whether a candidate (1, n) orbit is physical.

  `:velocity`: impact velocity for a candidate (1, n) orbit
  `:sigma`: the obstacle offset
  `:params`: parameters held fixed as the offset varies for a specified (1, n) orbit

  Returns `:true` if physical, `:false` if unphysical.
  """

  @spec is_physical?(float, float, OneNParams) :: boolean
  def is_physical?(velocity, _sigma, %OneNParams{} = _params) when velocity < 0 do
    # Negative velocity so can't be physical
    false
  end

  def is_physical?(velocity, sigma, %OneNParams{} = params) do
    if velocity == 0 and sigma > 0 do
      # Special case: periodic graze orbit. Assume physical
      true
    else
      # Verify numerically by computing next impact
      point = point_for_velocity(velocity, sigma, params)

      sys_params = %SystemParameters{omega: params.omega, r: params.r, sigma: sigma}

      # Should be periodic
      MotionBetweenImpacts.next_impact(point, sys_params) |> (& elem(&1, 0)).() |> (fn(next_point) ->
        cond do
          abs(point.v) < ImposcUtils.const_small() && next_point.v > ImposcUtils.const_small() -> false
          point.v != next_point.v and abs(point.v - next_point.v)/point.v > ImposcUtils.const_small() -> false
          abs(point.phi - next_point.phi) >  ImposcUtils.const_smallish() * params.period -> false
          true -> true
        end
      end).()
    end
  end
end

defmodule OneNLoci do
  @moduledoc """
  """

  def curves_for_fixed_omega(n, omega, r, num_points \\ 1000) do
    # Initialise parameters
    params = OneNParams.derive(omega, r, n)

    # Compute (1, n) velocities over range of offsets
    delta_s = 2 * params.sigma_s / num_points

    pairs = 0..num_points |> Stream.map(&(&1 * delta_s - params.sigma_s)) |>
      Stream.map(&({&1, OneNParams.velocities(&1, params)}))

    # Filter out unphysical (nullified) velocities
    filter_pairs = fn n -> (pairs |> Enum.map(&{elem(&1,0), elem(elem(&1,1), n)}) |> Enum.filter(&!is_nil(elem(&1,1)))) end

    0..1 |> Enum.map(&filter_pairs.(&1))
  end

  def vs(n, omega, r) do
    params = OneNParams.derive(omega, r, n)

    OneNParams.velocities(-params.sigma_s, params)
  end

  def orbits_for_params(%SystemParameters{} = params, n) do
    OneNParams.orbits(params.sigma, OneNParams.derive(params.omega, params.r, n))
  end
end

File: imposcpy/imposc/periodic.py
"""Classes and functions for analysing single-impact (1, n) periodic orbits, where n is the number of forcing cycles
between impacts.

An algebraic analysis reveals that closed-form expressions for (1, n) orbits can be obtained by solving a quadratic
equation for the impact velocity V_n. For a fixed forcing frequency `omega` and coefficient of restitution `r`,
the impact velocity Vn describes an ellipse in (`sigma`, V_n) space as the obstacle offset `sigma` is varied. Not all
points on the ellipse correspond to physical orbits. In particular, all orbits with Vn < 0 are obviously unphysical.
Further, as one traverses the ellipse anti-clockwise from the point where it intersects the `sigma`-axis for positive
`sigma`, then there is a point on the ellipse before one reaches the intersection with the `sigma`-axis for negative
`sigma`, at which orbits become unphysical - i.e. from an initial impact at (phi_n, V_n) the next impact is at time
before phi_n + n X forcing period. This point cannot be obtained in closed form and has to be computed numerically.

It can further be shown analytically that:

- where it corresponds to physical orbits, the upper branch of the ellipse is always stable and the lower branch is
always unstable
- as `omega` varies, the major axis of the ellipse rotates and it is vertical for `omega` = `n`
"""

from math import cos, sin, pi, sqrt, pow, acos
from typing import Tuple, Optional

from imposcpy.imposc.motion import ImpactPoint, phi, MotionBetweenImpacts
from imposcpy.imposc.parameters import SystemParameters
from imposcpy.imposc.constants import SMALL, SMALL_OMEGA, SMALLISH


class OneNParams:
    """ parameters for analysing single-impact (1, n) periodic orbits """

    def __init__(self, parameters: SystemParameters, n: int):
        """

        :param parameters: a `SystemParameters` instance defining the system (NOTE `parameters`.`sigma` is ignored
        because we are interested in generating the response when varying `sigma`)
        :param n: the number of forcing cycles between each impact
        """
        self._omega = parameters.omega
        self._r = parameters.r
        self._n = n

        # Derive coefficients used in the analysis
        self._r_minus = (1 - self._r) / self._omega

        self._period = 2 * pi * n / self._omega
        self._cn = cos(self._period)
        self._sn = sin(self._period)

        self._cs = 0

        if self._cs != 1:
            self._cs = self._sn * (1 + self._r) / (1 - self._cn)

        self._gamma = parameters.gamma()

        self._gamma2 = self._gamma * self._gamma

        self._phase_coeff = -self._r_minus / self._gamma / 2

        self._sigma_s = sqrt(self._gamma2 * (1 + pow(self._cs / self._r_minus, 2)))

    def discriminant(self, sigma: float) -> float:
        """ The discriminant of the quadratic equation for the impact velocity for a given value of the obstacle
        offset `sigma` """
        return 4 * (self._gamma2 * self._cs * self._cs - (sigma * sigma - self._gamma2) * self._r_minus * self._r_minus)

    def velocities(self, sigma: float) -> Tuple[Optional[float, None], Optional[float, None]]:
        """
        Solves the quadratic equation to return the velocities for candidate (1, n) orbits for a given obstacle
        offset. Depending on the value of the discriminant of the quadratic, there will be either zero, one (in the case
        of a double root) or two such velocities. Unphysical or negative-velocity orbits are not filtered out.

        :param sigma: the obstacle offset
        :return: A tuple comprising two velocities or two `None`s in the case of complex roots. In the case of a double
        root the two velocities will be the same.
        """
        discr = self.discriminant(sigma)

        if discr < 0:
            return None, None
        else:
            intercept = -2 * self._cs * sigma
            divisor = self._cs * self._cs + self._r_minus * self._r_minus

            vs = intercept / divisor

            d = sqrt(discr) / divisor

            return vs + d, vs - d

    def phase_for_velocity(self, v: float, sigma: float) -> Optional[float, None]:
        """
        Returns the phase corresponding to a solution of the quadratic equation for the velocity of a (1, n) orbit

        :param v: the candidate velocity
        :param sigma: the obstacle offset (as used in the quadratic)
        :return: the impact phase for the corresponding (1, n) orbit
        """
        if v is not None:
            arg = (sigma + self._cs * v / 2) / self._gamma

            angle = acos(arg)

            # Check for phase correction. sin(angle) should have opposite sign to v/gamma
            s = sin(angle)
            vg = v / self._gamma
            if (vg > 0 and s > 0) or (vg < 0 and s < 0):
                angle = 2 * pi - angle

            return phi(angle / self._omega, self._omega)
        else:
            return None

    def point_for_velocity(self, v: float, sigma: float) -> Optional[ImpactPoint, None]:
        """
        Returns in `ImpactPoint` corresponding to a solution of the quadratic equation for the velocity of a (1, n)
        orbit.

        :param v: the candidate velocity
        :param sigma: the obstacle offset (as used in the quadratic)
        :return: the point on the impact surface for the corresponding (1, n) orbit
        """
        if v is not None:
            return ImpactPoint(phi=self.phase_for_velocity(v, sigma), v=v)
        else:
            return None

    def orbits(self, sigma: float) -> Tuple[Optional[ImpactPoint, None], Optional[ImpactPoint, None]]:
        """
        Returns points on the impact surface corresponding to physical (1, n) orbits for a given obstacle offset. There
        will be either zero, one (in the case of a double root) or two such points.

        :param sigma: the obstacle offset
        :return: A two-place tuple comprising points on the impact surface corresponding to physical (1, n) orbits or
        `None`s in the case of unphysical orbits or complex roots to the quadratic.
        """
        return tuple([self.point_for_velocity(self.nullify_unphysical(v, sigma), sigma) for v in self.velocities(sigma)])

    def is_physical(self, v: float, sigma: float) -> bool:
        """
        Verifies the physicality of a candidate (1, n) orbit

        :param v: the candidate velocity
        :param sigma: the obstacle offset (as used in the quadratic)
        :return: `True` if the orbit is physical, `False` otherwise
        """
        if v is None or v <= 0:
            return False

        params = SystemParameters(omega=self._omega, r=self._r, sigma=sigma)

        start_point = self.point_for_velocity(v, sigma)

        motion = MotionBetweenImpacts(parameters=params, point=start_point)

        next_point = motion.next_impact()

        # Should be periodic

        return abs(next_point.v - start_point.v) / start_point.v < SMALL and abs(next_point.phi - start_point.phi) < \
               SMALLISH * self._period

    def nullify_unphysical(self, v: float, sigma: float) -> Optional[float, None]:
        """
        Filters out a candidate (1, n) orbit if it is unphysical

        :param v: the candidate velocity
        :param sigma: the obstacle offset (as used in the quadratic)
        :return: the candidate velocity if the orbit is physical, `None` otherwise
        """
        if self.is_physical(v, sigma):
            return v
        else:
            return None

    def __str__(self) -> str:
        """ Human-readable string representation """
        return f"{SMALL_OMEGA} = {self._omega}, r = {self._r}, n = {self._n}"

    def __repr__(self) -> str:
        """ An unambiguous string representation """
        return f"<{self.__class__.__name__}: {self.__str__()}>"

    @property
    def sigma_s(self) -> float:
        """ The positive intercept with the `sigma`-axis"""
        return self._sigma_s


def curves_for_fixed_omega(n: int, parameters: SystemParameters, num_points=1000):
    """
    Generates a `sigma`-response curve for a given forcing multiple and set of system parameters
    :param n: the number of forcing cycles between each impact
    :param parameters: a `SystemParameters` instance defining the system (NOTE `parameters`.`sigma` is ignored
        because we are interested in generating the response when varying `sigma`)
    :param num_points: the number of data points to record
    :return:
    """
    params = OneNParams(n=n, parameters=parameters)

    delta_s = 2 * params.sigma_s / (num_points - 1)

    sigmas = [i * delta_s - params.sigma_s for i in range(num_points)]

    pairs = [tuple([params.nullify_unphysical(v, sigma) for v in params.velocities(sigma)]) for sigma in sigmas]

    return sigmas, pairs, str(params)


if __name__ == '__main__':
    sys_params = SystemParameters(omega=2,r=0.8, sigma=0)

    one_n_params = OneNParams(parameters=sys_params, n=1)

    num_points = 10

    delta_s = 2 * one_n_params.sigma_s / (num_points - 1)

    sigmas = [i * delta_s - one_n_params.sigma_s for i in range(num_points)]

    pairs = [(sigma, one_n_params.velocities(sigma)) for sigma in sigmas]

    [print(pair, one_n_params.phase_for_velocity(pair[1][0], pair[0])) for pair in pairs]